lib/nrf_modem_lib/nrf_modem_lib_trace.c

```diff
@@ -136,27 +152,53 @@ static void bps_log(struct k_work *item)
 
 	trace_bytes_received = 0;
 
+	LOG_INF("Written: %d, read: %d", trace_bytes_received_total, trace_bytes_read_total);
 	LOG_INF("Trace bitrate (bps): %u", trace_data_bps_avg);
 
 	k_work_schedule(&bps_log_work, BPS_LOG_PERIOD);
 }
 
 #define UPDATE_TRACE_BYTES_RECEIVED(frags, n_frags) update_trace_bytes_received(frags, n_frags)
+#define UPDATE_TRACE_BYTES_READ(bytes) update_trace_bytes_read(bytes)
 #else
 #define UPDATE_TRACE_BYTES_RECEIVED(...)
+#define UPDATE_TRACE_BYTES_READ(...)
 #endif
 
+__weak void nrf_modem_lib_trace_callback(enum nrf_modem_lib_trace_event evt)
+{
+	LOG_ERR("%s was called with event %d but no callback is set", __func__, evt);
+}
+
+void nrf_modem_lib_trace_callback_set(nrf_modem_lib_trace_callback_t callback)
+{
+	if (!callback) {
+		trace_callback = nrf_modem_lib_trace_callback;
+		return;
+	}
+
+	trace_callback = callback;
+}
```

Comment:
Trying to remember how this was supposed to work. I think we wanted to handle the scenario where `_NOSPACE_SIGNAL` is selected but the user does not set that callback (or sets it after initializing the library).

The callback has to be set before `nrf_modem_lib_init`, because the trace thread will start automatically when the library is started, and the callback must be ready to receive events by then.

If the user controls when `nrf_modem_lib_init` is called, we have to document `nrf_modem_lib_trace_callback_set` must be called before `nrf_modem_lib_init`. We could enforce this at runtime by checking in `nrf_modem_lib_init` that a callback is set in `nrf_modem_trace.c` and return an error otherwise.

If the user does not control when `nrf_modem_lib_init` is called we would have to ensure a callback exists at the earliest time possible (we can assume `nrf_modem_lib_init` is called during `SYS_INIT` in this case).

I think checking  that `nrf_modem_lib_trace_callback_set` has been called during `nrf_modem_lib_init` gives little possibility to the application to do that in case `nrf_modem_lib_init` is called during `SYS_INIT`, unless the application also setups a `SYS_INIT` function that calls `nrf_modem_lib_trace_callback_set`. That seems like a workaround :)

An alternative that would work in both cases would be to remove `nrf_modem_lib_trace_callback_set` and the weak implementation of `nrf_modem_lib_trace_callback`. Document that if `_NOSPACE_SIGNAL` is selected, the user must implement a function called `nrf_modem_lib_trace_callback` themselves. In the code, we will need to do something like this:

```
#ifdef _NOSPACE_SIGNAL 
extern __weak void nrf_modem_lib_trace_callback(...);
BUILD_ASSERT(nrf_modem_lib_trace_callback, "No nrf_modem_lib_trace_callback() implementation");
#endif

/* when you need to invoke the callback */
#ifdef _NOSPACE_SIGNAL
nrf_modem_lib_trace_callback(event);
#endif
```

Reply:
Discussed offline. Will remove the `_set` function and fail when initializing the flash backend with `_SIGNAL` if no callback is defined. We cannot use BUILD_ASSERT for this as the callback is kept as a symbol until linking. Will add some documentation for this.